src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,877 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for finding affected ranges from over window range cache for a
+//! given set of changes (delta).
+
+use std::ops::Bound;
+
+use delta_btree_map::DeltaBTreeMap;
+use risingwave_common::row::OwnedRow;
+use risingwave_expr::window_function::{FrameBound, RowsFrameBounds};
+
+use super::over_partition::CacheKey;
+
+// -------------------------- ↓ PUBLIC INTERFACE ↓ --------------------------
+
+/// Merge several `ROWS` frames into one super frame. The returned super frame is
+/// guaranteed to be *canonical*, which means that the `CURRENT ROW` is always
+/// included in the returned frame.
+pub(super) fn merge_rows_frames(rows_frames: &[&RowsFrameBounds]) -> RowsFrameBounds {
+    if rows_frames.is_empty() {
+        // this can simplify our implementation
```

Comment:
Could you please more docs on correctness here?

Reply:
```suggestion
        // this can simplify our implementation because ...
```

Reply:
Added